Source/WebCore/crypto/CryptoKeyData.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "CryptoAlgorithmIdentifier.h"
+#include "CryptoKeyType.h"
+#include "CryptoKeyUsage.h"
+#include "JsonWebKey.h"
+
+namespace WebCore {
+
+enum class CryptoKeyClass : uint8_t {
+    AES,
+    EC,
+    HMAC,
+    OKP,
+    RSA,
+    Raw
+};
+
+struct CryptoKeyData {
+    CryptoKeyData(CryptoKeyClass keyClass, CryptoAlgorithmIdentifier algorithmIdentifier, bool extractable, CryptoKeyUsageBitmap usages, std::optional<Vector<uint8_t>> key, std::optional<JsonWebKey> jwk = std::nullopt, std::optional<CryptoAlgorithmIdentifier> hashAlgorithmIdentifier = std::nullopt, std::optional<String>&& namedCurveString = std::nullopt, std::optional<size_t> lengthBits = std::nullopt, std::optional<CryptoKeyType> type = std::nullopt)
```

Comment:
Why do you need both jwk and Vector  together.. wouldn't jwk work everywhere ?
and if it's either then std::variant<jwk,Vector> seems like a natural choice.

Reply:
CryptoKeyType also.

Reply:
> Why do you need both jwk and Vector together.. wouldn't jwk work everywhere ?
`CryptoKeyRaw` cannot export jwk

> The algorithm Identifier should probably also not really need optional ? it should always be there.

Algorithm identifier is not optional, hash algorithm identifier is; see members of `CryptoKeyData` starting at line 73 of this file.

> CryptoKeyType also.

`CryptoKeyType` is optional because some `importJwk` functions do not require that.
When a parameter is not required for all functions, I make it optional.

> Or is it something to do with IPC serialization that all items are optional ?

The logic is we extract `CryptoKeyData` from `CryptoKey` (with `exportJwk` and other property accessor function) on the web process side, send `CryptoKeyData` to UI process, and recreate `CryptoKey` from `CryptoKeyData` (with `CryptoKey::create(CryptoKeyData&&)`)on the UI process side.